Ambiguous JPMS module reference in maven projects

In maven projects, published "umbrella" artifacts (like "kotlinx-serialization-json") works as a gateway to corresponding `-jvm` artifacts. The former do not have a JPMS module bundled inside, so the module name is inferred from the artifact name. As a result, there are two artifacts providing the same module:
- `kotlinx-serialization-json` provides `kotlinx.serialization.json` as it was inferred from the module name.
- `kotlinx-serialization-json-jvm` provides `kotlinx.serialization.json`, well, because it actually provides it.

**To Reproduce**

Here's a reproducer: https://github.com/fzhinkin/kotlinx-serialization-jpms-conflict
```
$ mvm compile
...
[WARNING] /Users/filipp.zhinkin/.m2/repository/org/jetbrains/kotlinx/kotlinx-serialization-json-jvm/1.8.0/kotlinx-serialization-json-jvm-1.8.0.jar: (-1, -1) The root is ignored because a module with the same name 'kotlinx.serialization.json' has been found earlier on the module path at: /Users/filipp.zhinkin/.m2/repository/org/jetbrains/kotlinx/kotlinx-serialization-json/1.8.0/kotlinx-serialization-json-1.8.0.jar
[ERROR] /Users/filipp.zhinkin/Development/kotlinx-serialization-jmps-conflict/src/main/kotlin/Dummy.kt: (4, 30) Unresolved reference 'json'.
[ERROR] /Users/filipp.zhinkin/Development/kotlinx-serialization-jmps-conflict/src/main/kotlin/Dummy.kt: (10, 33) Unresolved reference 'Json'.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
```

**Expected behavior**

"Umbrella" artifacts like `kotlinx-serialization-core` or `kotlinx-serialization-json` should provide some JPMS modules that won't clash with modules provided by actual `***-jvm` artifacts.

--- a/buildSrc/src/main/kotlin/Java9Modularity.kt
+++ b/buildSrc/src/main/kotlin/Java9Modularity.kt
@@ -241,4 +241,28 @@ object Java9Modularity {
             )
         })
     }
+
+    /**
+     * For multiplatform Gradle modules, generate and set `Automatic-Module-Name` in metadata JAR's manifest.
+     *
+     * Generated automatic (JPMS) module name has the following format:
+     * `<Gradle module name>. artifact_disambiguating_module`.
+     *
+     * For multiplatform projects, a metadata artifact is the one without a platform-specific suffix,
+     * and it always depends on corresponding `-jvm` artifact (for convenience on build systems other than Gradle).
+     * For a JVM project depending on such an artifact and using JPMS, it may result in the automatic module name clash
+     * with a module provided by the `-jvm` artifact. By explicitly setting a non-clashing automatic module name in
+     * metadata JAR's manifest, we're mitigating this issue.
+     */
+    fun Project.configureMetadataJarAutomaticModuleName() {
+        val kotlin = extensions.findByType<KotlinMultiplatformExtension>() ?: return
+        val moduleName = project.name.replace('-', '.') + ".artifact_disambiguating_module"
+        tasks.withType<Jar>().named(kotlin.metadata().artifactsTaskName) {
+            manifest {
+                attributes(
+                    "Automatic-Module-Name" to moduleName,
+                )
+            }
+        }
+    }
 }

--- a/core/build.gradle.kts
+++ b/core/build.gradle.kts
@@ -1,4 +1,5 @@
 import Java9Modularity.configureJava9ModuleInfo
+import Java9Modularity.configureMetadataJarAutomaticModuleName
 import org.jetbrains.kotlin.gradle.targets.js.ir.*
 
 /*
@@ -69,6 +70,7 @@ tasks.withType<Jar>().named(kotlin.jvm().artifactsTaskName) {
 }
 
 configureJava9ModuleInfo()
+configureMetadataJarAutomaticModuleName()
 
 tasks.withType<KotlinJsIrLink>().configureEach {
     compilerOptions.freeCompilerArgs.add("-Xwasm-enable-array-range-checks")

--- a/formats/cbor/build.gradle.kts
+++ b/formats/cbor/build.gradle.kts
@@ -1,4 +1,5 @@
 import Java9Modularity.configureJava9ModuleInfo
+import Java9Modularity.configureMetadataJarAutomaticModuleName
 
 /*
  * Copyright 2017-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
@@ -36,3 +37,4 @@ kotlin {
 }
 
 configureJava9ModuleInfo()
+configureMetadataJarAutomaticModuleName()

--- a/formats/json-io/build.gradle.kts
+++ b/formats/json-io/build.gradle.kts
@@ -2,8 +2,7 @@
  * Copyright 2017-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
  */
 import Java9Modularity.configureJava9ModuleInfo
-import org.jetbrains.dokka.gradle.*
-import java.net.*
+import Java9Modularity.configureMetadataJarAutomaticModuleName
 
 plugins {
     kotlin("multiplatform")
@@ -32,6 +31,7 @@ kotlin {
 }
 
 project.configureJava9ModuleInfo()
+project.configureMetadataJarAutomaticModuleName()
 
 dokka.dokkaSourceSets.configureEach {
     externalDocumentationLinks.register("kotlinx-io") {

--- a/formats/json-okio/build.gradle.kts
+++ b/formats/json-okio/build.gradle.kts
@@ -2,8 +2,7 @@
  * Copyright 2017-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
  */
 import Java9Modularity.configureJava9ModuleInfo
-import org.jetbrains.dokka.gradle.*
-import java.net.*
+import Java9Modularity.configureMetadataJarAutomaticModuleName
 
 plugins {
     kotlin("multiplatform")
@@ -33,6 +32,7 @@ kotlin {
 }
 
 project.configureJava9ModuleInfo()
+project.configureMetadataJarAutomaticModuleName()
 
 dokka.dokkaSourceSets.configureEach {
     externalDocumentationLinks.register("okio") {

--- a/formats/json/build.gradle.kts
+++ b/formats/json/build.gradle.kts
@@ -1,5 +1,5 @@
 import Java9Modularity.configureJava9ModuleInfo
-import org.jetbrains.kotlin.gradle.tasks.*
+import Java9Modularity.configureMetadataJarAutomaticModuleName
 
 /*
  * Copyright 2017-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
@@ -53,3 +53,4 @@ kotlin {
 }
 
 configureJava9ModuleInfo()
+configureMetadataJarAutomaticModuleName()

--- a/formats/properties/build.gradle.kts
+++ b/formats/properties/build.gradle.kts
@@ -1,4 +1,5 @@
 import Java9Modularity.configureJava9ModuleInfo
+import Java9Modularity.configureMetadataJarAutomaticModuleName
 
 /*
  * Copyright 2017-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
@@ -36,3 +37,4 @@ kotlin {
 }
 
 configureJava9ModuleInfo()
+configureMetadataJarAutomaticModuleName()

--- a/formats/protobuf/build.gradle.kts
+++ b/formats/protobuf/build.gradle.kts
@@ -2,6 +2,7 @@
  * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
  */
 import Java9Modularity.configureJava9ModuleInfo
+import Java9Modularity.configureMetadataJarAutomaticModuleName
 
 plugins {
     kotlin("multiplatform")
@@ -34,3 +35,4 @@ kotlin {
 }
 
 configureJava9ModuleInfo()
+configureMetadataJarAutomaticModuleName()

